### PR TITLE
Shared ptr events

### LINKED ISSUE #477: [Level Zero] Unit_hipMemcpy2D_H2D-D2D-D2H - float
dGPU L0
```
hipMemcpy2D: /home/pvelesko/actions-runner/_work/chip-spv/chip-spv/src/backend/Level0/CHIPBackendLevel0.cc:239: ze_event_handle_t CHIPEventLevel0::peek(): Assertion `!Deleted_ && "Event use after delete!"' failed.
Filters: Unit_hipMemcpy2D_H2D-D2D-D2H - float
```

### PR DESCRIPTION
**Pull request by pvelesko**
This PR gets rid of our own reference counting and overall event management.
* `CHIPBackend::createCHIPEvent` pure virtual method now expects a return of a `std::shared_ptr<CHIPEvent>
* * There is no option for returning a non-shared event.
* User events are stored in a new vector: `CHIPBackend::UserEvents` - since all events are created as shared_ptr, destroying these events involves taking the raw pointer and looking it up in `CHIPBackend::UserEvents` 

Fixes #477

### CODE REVIEW

File: src/backend/Level0/CHIPBackendLevel0.cc [hunk 1]
```diff
@@ -1296,35 +1257,36 @@ CHIPEvent *CHIPQueueLevel0::enqueueMarkerImpl() {
...
 }
 
-CHIPEvent *
-CHIPQueueLevel0::enqueueBarrierImpl(std::vector<CHIPEvent *> *EventsToWaitFor) {
-  CHIPEventLevel0 *EventToSignal =
-      (CHIPEventLevel0 *)Backend->createCHIPEvent(ChipContext_);
-
-  EventToSignal->sanityCheck();
+std::shared_ptr<CHIPEvent> CHIPQueueLevel0::enqueueBarrierImpl(
+    std::vector<std::shared_ptr<CHIPEvent>> EventsToWaitFor) {
```
Comment (franz) [inline src/backend/Level0/CHIPBackendLevel0.cc:1268]: any reason to not pass the vector argument by a (const) reference ?

File: src/backend/Level0/CHIPBackendLevel0.cc [hunk 2]
```diff
@@ -1296,35 +1257,36 @@ CHIPEvent *CHIPQueueLevel0::enqueueMarkerImpl() {
...
+  SignalEventHandle =
+      std::static_pointer_cast<CHIPEventLevel0>(EventToSignal)->peek();
 
   if (NumEventsToWaitFor > 0) {
     EventHandles = new ze_event_handle_t[NumEventsToWaitFor];
     for (size_t i = 0; i < NumEventsToWaitFor; i++) {
-      CHIPEventLevel0 *ChipEventLz = (CHIPEventLevel0 *)(*EventsToWaitFor)[i];
+      std::shared_ptr<CHIPEvent> ChipEvent = EventsToWaitFor[i];
+      std::shared_ptr<CHIPEventLevel0> ChipEventLz =
+          std::dynamic_pointer_cast<CHIPEventLevel0>(ChipEvent);
```
Comment (franz) [inline src/backend/Level0/CHIPBackendLevel0.cc:1287]: Unless i'm mistaken, this will introduce requirement to compile the code with RTTI.

Comment (pvelesko) [reply src/backend/Level0/CHIPBackendLevel0.cc:1287]: typo

Comment (pvelesko) [reply src/backend/Level0/CHIPBackendLevel0.cc:1287]: typo. Fixed.

**Pull request merged**